File: GNB/gnb.c
#include <math.h>
#include <stdint.h>

#ifdef USE_CMSIS
#include "arm_math.h"
#else

#include "gnb.h"

void arm_max_f32(float32_t *p, uint32_t size, float32_t *max, uint32_t *index)
{
    *max=p[0];
    *index=0;
    for(uint32_t n=1; n<size; n++)
    {
        if (p[n]>*max)
        {
            *max=p[n];
            *index=n;
        }
    }
    
}

//                    p(y)*p(x1|y)p(x2|y)...
// p(y|x1,x2,...)=  --------------------------
//                        p(x1)p(x2)...
// 这里计算概率的对数
// S->numberOfClasses   y的离散取值个数
// S->vectorDimension   x观测量个数
// S->classPriors       先验p(y)
// S->theta和S->sigma   分别是每个Y的取值和每个x观测量对应的高斯模型参数
//                      分段存放，每个y的取值一段(每段长x的个数）
// in                   观测向量，就是x1, x2, ...
// pBuffer              存放y的每个类别的概率对数
uint32_t arm_gaussian_naive_bayes_predict_f32(
    const arm_gaussian_naive_bayes_instance_f32 *S, // 数据结构
    const float32_t * in, 
    float32_t *pBuffer)
{
    uint32_t nbClass;
    uint32_t nbDim;
    const float32_t *pPrior = S->classPriors;
    const float32_t *pTheta = S->theta;
    const float32_t *pSigma = S->sigma;
    float32_t *buffer = pBuffer;
    const float32_t *pIn=in;
    float32_t result;
    float32_t sigma;
    float32_t tmp;
    float32_t acc1,acc2;
    uint32_t index;

    pTheta=S->theta;
    pSigma=S->sigma;

    for(nbClass = 0; nbClass < S->numberOfClasses; nbClass++)   // 遍历每个Y的类别
    {
        pIn  = in;          // 恢复pIn指针（指向各个观测量）
        tmp  = 0.0;
        acc1 = 0.0f;
        acc2 = 0.0f;
        for(nbDim = 0; nbDim < S->vectorDimension; nbDim++)     // 遍历每个观测量（假设每个是独立高斯）
        {
           sigma = *pSigma + S->epsilon;
           acc1 += logf(2.0f * PI_F * sigma);                   // 对数相加用于实现相乘
           acc2 += (*pIn - *pTheta) * (*pIn - *pTheta) / sigma; // exp相乘对应指数相加

           pIn++;
           pTheta++;        // 每个观测量序号和每个Y的离散取值（类别）对应一个模型
           pSigma++;        // 同上
        }

        tmp = -0.5f * acc1; // 作用在对数上，对应开根号倒数
        tmp -= 0.5f * acc2; // 减号是因为exp指数上是负的，前面计算acc2的时候没有取负号，这里补上

        *buffer = tmp + logf(*pPrior++);                        // 对数相加对应乘以先验
        buffer++;
    }

    arm_max_f32(pBuffer,S->numberOfClasses,&result,&index);

    return(index);
}

#endif

File: GNB/gnb.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# 高斯朴素贝叶斯模型

from sklearn import datasets
iris = datasets.load_iris()

from sklearn.naive_bayes import GaussianNB
model = GaussianNB()
model.fit(iris.data, iris.target)
y_pred = model.predict(iris.data)

print('[INF] test error: %d/%d'%
      ((iris.target != y_pred).sum(),iris.data.shape[0]))

#num_cls, num_dim = model.sigma_.shape
num_cls, num_dim = model.var_.shape
num_dat=len(iris['target'])

import platform,os,sys
# 设置当前运行目录
os.chdir(sys.path[0])

# 生成C头文件
print('[INF] generating C header...')
fp=open('export_code/gnb_test.h','wt')
fp.write('#ifndef __GNB_TEST_H__\n')
fp.write('#define __GNB_TEST_H__\n')
fp.write('\n')
fp.write('#ifdef USE_CMSIS\n')
fp.write('#include "arm_math.h"\n')
fp.write('#else\n')
fp.write('#include "../gnb.h"\n')
fp.write('#endif\n')
fp.write('\n')
fp.write('#define NUM_CLS %d\n'%num_cls)
fp.write('#define NUM_DIM %d\n'%num_dim)
fp.write('#define NUM_DAT %d\n'%num_dat)
fp.write('#endif\n')
fp.close()

## 生成C源代码
print('[INF] generating C source...')
fp=open('export_code/gnb_test.c','wt')
fp.write('#ifdef USE_CMSIS\n')
fp.write('#include "arm_math.h"\n')
fp.write('#else\n')
fp.write('#include "../gnb.h"\n')
fp.write('#endif\n')
fp.write('\n')
fp.write('#include "gnb_test.h"\n')
fp.write('\n')
fp.write('const float32_t model_theta[NUM_CLS*NUM_DIM] = \n{')
for t_ in model.theta_:
    fp.write('\n    ')
    for t in t_: fp.write('%e, '%t)
fp.write('\n};\n')   

fp.write('const float32_t model_sigma[NUM_CLS*NUM_DIM] = \n{')
for s_ in model.var_:
    fp.write('\n    ')
    for s in s_: fp.write('%e, '%s)
fp.write('\n};\n')    

fp.write('const float32_t model_priors[NUM_CLS] = \n{\n    ')
if model.class_prior_ is not None:
    for p in model.class_prior_: fp.write('%e, '%p)
else:
    for _ in range(num_cls): fp.write('%e, '%(1.0/float(num_cls)))
fp.write('\n};\n')
fp.write('\n')

fp.write('const float32_t test_in[NUM_DAT*NUM_DIM] = \n{')
for d_ in iris['data']:
    fp.write('\n    ')
    for d in d_: fp.write('%e, '%d)
fp.write('\n};\n') 
fp.write('const uint32_t test_out[] = \n{')
for n,t in enumerate(iris['target']):
    if n%16==0: fp.write('\n    ')
    fp.write('%d, '%t)
fp.write('\n};\n')

fp.write('\n')
fp.write('int32_t gnb_test()\n')
fp.write('{\n')
fp.write('    uint32_t err=0;\n')
fp.write('    float32_t prob[NUM_CLS];\n')
fp.write('\n')
fp.write('    arm_gaussian_naive_bayes_instance_f32 model;\n')
fp.write('    model.vectorDimension=NUM_DIM;\n')
fp.write('    model.numberOfClasses=NUM_CLS;\n')              
fp.write('    model.theta=model_theta;\n')
fp.write('    model.sigma=model_sigma;\n')
fp.write('    model.classPriors=model_priors;\n')

fp.write('    model.epsilon=0.0;\n')
#fp.write('    model.epsilon=%e;\n'%model.epsilon_)

fp.write('\n')
fp.write('    err=0;\n')
fp.write('    for (int n=0; n<NUM_DAT; n++)\n')
fp.write('        if (test_out[n]!=arm_gaussian_naive_bayes_predict_f32(&model,test_in+n*NUM_DIM,prob))\n')
fp.write('            err++;\n')
fp.write('    return err;\n')
fp.write('}\n')

fp.write('\n')
fp.write('#include <stdio.h>\n')
fp.write('int main()\n')
fp.write('{\n')
fp.write('    int32_t num_err=gnb_test();\n')
fp.write('    printf("[INF] test error: %d/%d\\n",num_err,NUM_DAT);\n')
fp.write('}\n')

fp.close()


